src/algebra/lie/free.lean

```diff
@@ -0,0 +1,229 @@
+  sub            := λ x y, x + ((-1 : R) • y),
+  sub_eq_add_neg := λ x y, rfl,
+  add_left_neg   :=
+    by { rintros ⟨a⟩, change quot.mk _ _ = _, erw [neg_smul, one_smul, add_left_neg], refl, },
+  ..(infer_instance : has_add _)}
+
+/-- Note that here we turn the `has_mul` coming from the `non_unital_non_assoc_semiring` structure
+on `lib R X` into a `has_bracket` on `free_lie_algebra`. -/
+instance : has_bracket (free_lie_algebra R X) (free_lie_algebra R X) :=
+{ bracket := quot.map₂ (*) (λ a b c, rel.mul_left a b c) (λ a b c, rel.mul_right a b c), }
+
+instance : lie_ring (free_lie_algebra R X) :=
+{ add_lie     := by { rintros ⟨a⟩ ⟨b⟩ ⟨c⟩, change quot.mk _ _ = quot.mk _ _, rw add_mul, },
+  lie_add     := by { rintros ⟨a⟩ ⟨b⟩ ⟨c⟩, change quot.mk _ _ = quot.mk _ _, rw mul_add, },
+  lie_self    := by { rintros ⟨a⟩, exact quot.sound (rel.lie_self a), },
+  leibniz_lie := by { rintros ⟨a⟩ ⟨b⟩ ⟨c⟩, exact quot.sound (rel.leibniz_lie a b c), }, }
+
+instance : module R (free_lie_algebra R X) :=
+{ one_smul  := by { rintros ⟨a⟩, change quot.mk _ _ = quot.mk _ _, rw one_smul, },
+  mul_smul  := by { rintros t₁ t₂ ⟨a⟩, change quot.mk _ _ = quot.mk _ _, rw mul_smul, },
+  add_smul  := by { rintros t₁ t₂ ⟨a⟩, change quot.mk _ _ = quot.mk _ _, rw add_smul, },
+  smul_add  := by { rintros t ⟨a⟩ ⟨b⟩, change quot.mk _ _ = quot.mk _ _, rw smul_add, },
+  zero_smul := by { rintros ⟨a⟩, change quot.mk _ _ = quot.mk _ _, rw zero_smul, },
+  smul_zero := λ t, by { change quot.mk _ _ = quot.mk _ _, rw smul_zero, }, }
+
+instance : lie_algebra R (free_lie_algebra R X) :=
+{ lie_smul :=
+  begin
+    rintros t ⟨a⟩ ⟨c⟩,
+    change quot.mk _ (a • (t • c)) = quot.mk _ (t • (a • c)),
+    rw ← smul_comm,
+  end, }
+
+variables {X}
+
+/-- The embedding of `X` into the free Lie algebra of `X` with coefficients in the commutative ring
+`R`. -/
+def of : X → free_lie_algebra R X := λ x, quot.mk _ (lib.of R x)
+
+variables {L : Type w} [lie_ring L] [lie_algebra R L]
+
+local attribute [instance] lie_ring.to_non_unital_non_assoc_semiring
+local attribute [instance] lie_algebra.is_scalar_tower
+local attribute [instance] lie_algebra.smul_comm_class
+
+/-- An auxiliary definition used to construct the equivalence `lift` below. -/
+def lift_aux (f : X → L) := lib.lift R f
+
+lemma lift_aux_map_smul (f : X → L) (t : R) (a : lib R X) :
+  lift_aux R f (t • a) = t • lift_aux R f a :=
+non_unital_alg_hom.map_smul _ t a
+
+lemma lift_aux_map_add (f : X → L) (a b : lib R X) :
+  lift_aux R f (a + b) = (lift_aux R f a) + (lift_aux R f b) :=
+non_unital_alg_hom.map_add _ a b
+
+lemma lift_aux_map_mul (f : X → L) (a b : lib R X) :
+  lift_aux R f (a * b) = ⁅lift_aux R f a, lift_aux R f b⁆ :=
+non_unital_alg_hom.map_mul _ a b
+
+lemma lift_aux_spec (f : X → L) (a b : lib R X) (h : free_lie_algebra.rel R X a b) :
+  lift_aux R f a = lift_aux R f b :=
+begin
+  induction h with c a' b' c' t a' b' h₁ h₂ a' b' c' h₁ h₂ a' b' c' h₁ h₂ a' b' c' h₁ h₂,
+  { -- rel.lie_self
```

Comment:
I don't have experience with it, but I think you can turn these comments into code using `case`. It also allows you to introduce the variable names per case (instead of the massive `induction`-line above).

Reply:
Oh much better, thanks!